### ISSUE TITLE
Handling of nullable-aware comparisons with `None` first

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#2201]
## Describe the bug

Comparisons between a nullable class instance and a corresponding non-nullable instance fail, despite the comparison methods all supporting comparisons with `None`.

## To reproduce

```python
import operator
from typing import Callable

class Cmp:

    def __init__(self, value: int):
        self.value = value

    def _cmp(self, other: Any, op: Callable[[Any, Any], Any]):
        if other is None:
            return False
        elif not isinstance(other, Cmp):
            return NotImplemented
        else:
            return op(self.value, other.value)

    def __eq__(self, other: Any):
        return self._cmp(other, operator.eq)

    def __lt__(self, other: Any):
        return self._cmp(other, operator.lt)

    # ...the other four operators...
```
```python
def valid(value: Optional[Cmp], needed: Optional[Cmp]):
    if needed is None:
        return True
    else:
        return value >= needed
        #      ^^^^^
        # Operator ">=" not supported for "None"
```

## Expected behavior

The null check on `needed` should be sufficient to avoid a `None`-`None` comparison.  This succeeds at runtime if `value` is `None` by implicitly reversing the comparison (effectively calling `needed.__lt__(value)`).

I assume this is down to `object.__lt__` and friends typed to accept `object` for their `other` parameter (https://github.com/microsoft/pyright/discussions/1411#discussioncomment-307085), despite the use of `Any` here.  Possibly related to #1676 with respect to widening the argument type?

There is of course the obvious workaround of reversing the values myself, but there's a readability gain on having it this way around (putting the comparison value first has the same "Yoda notation" semantics as e.g. `if 1 == value:`).

## VS Code extension or command-line

Pyright 1.1.162 in VSCode.

**Comment by erictraut** [comment]
The `reportOptionalOperand` diagnostic check looks for cases where the LHS operand is a union type that includes `None`. It reports this as an error condition and then removes the `None` from the union and proceeds with the resulting type. This approach deals well with a very common case in Python code and allows users to disable the diagnostic if they find that it's too noisy.

I agree that in your particular code above, the error is a false positive. I'll need to think more about whether there's a way to preserve the current behavior of the `reportOptionalOperand` while eliminating this false positive. If they're in conflict, we'll need to decide which is more important — to retain the current behavior of this diagnostic check or fix this edge-case false positive.

**Comment by erictraut** [comment]
I found a way to preserve the existing behavior of `reportOptionalOperand` while eliminating this false positive. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.165, which I just published. It will also be included in the next release of pylance.
